lib/vmpooler/metrics/promstats.rb

```diff
@@ -0,0 +1,306 @@
+      @p_metrics.each do |_name, metric_spec|
+        if metric_spec.key? :metric_suffixes
+          # Iterate thru the suffixes if provided to register multiple counters here.
+          metric_spec[:metric_suffixes].each do |metric_suffix|
+            add_prometheus_metric(
+              metric_spec,
+              "#{metric_spec[:prom_metric_prefix]}_#{metric_suffix[0]}",
+              "#{metric_spec[:docstring]} #{metric_suffix[1]}"
+            )
+          end
+        else
+          # No Additional counter suffixes so register this as metric.
+          add_prometheus_metric(
+            metric_spec,
+            metric_spec[:prom_metric_prefix],
+            metric_spec[:docstring]
+          )
+        end
+      end
+    end
+
+    # locate a metric and check/interpet the sub-fields.
+    def find_metric(label)
+      sublabels = label.split('.')
+      metric_key = sublabels.shift.to_sym
+      raise("Invalid Metric #{metric_key} for #{label}") unless @p_metrics.key? metric_key
+
+      metric = @p_metrics[metric_key]
+
+      if metric.key? :metric_suffixes
+        metric_subkey = sublabels.shift.to_sym
+        raise("Invalid Metric #{metric_key}_#{metric_subkey} for #{label}") unless metric[:metric_suffixes].key? metric_subkey.to_sym
+
+        metric[:metric_name] = "#{metric[:prom_metric_prefix]}_#{metric_subkey}"
+      else
+        metric[:metric_name] = metric[:prom_metric_prefix]
+      end
+
+      # Check if we are looking for a parameter value at last element.
+      if metric.key? :param_labels
+        metric[:labels] = {}
+        # Special case processing here - if there is only one parameter label then make sure
+        # we append all of the remaining contents of the metric with "." separators to ensure
+        # we get full nodenames (e.g. for Migration to node operations)
+        if metric[:param_labels].length == 1
+          metric[:labels][metric[:param_labels].first] = sublabels.join('.')
+        else
+          metric[:param_labels].reverse_each do |param_label|
+            metric[:labels][param_label] = sublabels.pop(1).first
+          end
+        end
+      end
+      metric
+    end
+
+    # Helper to get lab metrics.
+    def get(label)
+      metric = find_metric(label)
+      [metric, @prometheus.get(metric[:metric_name])]
+    end
+
+    # Note - Catch and log metrics failures so they can be noted, but don't interrupt vmpooler operation.
+    def increment(label)
+      begin
+        @mutex.synchronize do
+          counter_metric, c = get(label)
+          c.increment(labels: counter_metric[:labels])
+        end
+      rescue StandardError => e
+        @logger.log('s', "[!] prometheus error logging metric #{label} increment : #{e}")
+      end
+    end
+
+    def gauge(label, value)
+      begin
+        unless value.nil?
+          @mutex.synchronize do
+            gauge_metric, g = get(label)
+            g.set(value.to_i, labels: gauge_metric[:labels])
```

Comment:
Funny thing about putting this inside a mutex is that the client library appears to also be using a mutex.  https://github.com/prometheus/client_ruby/blob/master/lib/prometheus/client/data_stores/synchronized.rb#L37

I also notice that the last commit to that file... removed `concurrent-ruby` as a dependency. It makes me wonder if there's something about the client itself that isn't actually thread-safe. https://github.com/prometheus/client_ruby/commit/640194536647081cdca213609cb21c02bbfb0574